### DISCUSSION TITLE
What's `$DOCKER_RAMDISK` and is it still a thing?

### DISCUSSION THREAD

**Discussion by stephan-henningsen** [moby/moby#47035] [Q&A] [answered]
Hi,

1. Is `$DOCKER_RAMDISK` still supported by Docker today?
2. And if so, what exactly does it do?

I'm managing a small in-memory Linux distro exclusive for running Docker.
In that regard I've come across the environment variable `DOCKER_RAMDISK`
in similar projects and older blog posts.  But I haven't been able to find any
contemporary, official documentation on this feature.

I did find two github issues on the subject, but they only made me more confused:

- https://github.com/moby/moby/issues/21631
- https://github.com/moby/moby/pull/21679

I tried a rudimentary search in the source code, and I did find the
`$DOCKER_RAMDISK` to be read, but my go-fu isn't strong enough to
determine if or how it's actually used;

- https://github.com/search?q=repo%3Amoby%2Fmoby%20DOCKER_RAMDISK&type=code

Any help to shed light on this subject would be much appreciated.

--Stephan

**Comment by AkihiroSuda** [comment]
> 1. Is $DOCKER_RAMDISK still supported by Docker today?

It is "supported" in the sense that it is still implemented, but I don't think this configuration is well tested today.

> 2. And if so, what exactly does it do?

It runs `runc` with `--no-pivot` to disable `pivot_root(2)`, which is often unavailable on RAM disk environments.

https://github.com/opencontainers/runc/blob/v1.1.11/libcontainer/rootfs_linux.go#L138-L147

This may potentially decrease security.

**Comment by stephan-henningsen** [comment]
Thanks, good to know.

Can you perhaps elaborate on what the `pivot_root` is actually used for?  Is it a security measure?  Is it `runc` that uses some directory specific to a given container (maybe`/var/lib/docker/containers/$container_id) in the event someone should manage to break out of it?  What is its actual intent and effect?

**Reply by stephan-henningsen** [reply]
Or put differently:
```
$ runc run --help | grep -- --no-pivot
   --no-pivot          not use pivot root to jail process inside rootfs.  This should be used whenever the rootfs is on top of a ramdisk
```
Why?

**Reply by AkihiroSuda** [reply]
In a nutshell, pivot_root is a more secure version of chroot.
It is less likely to be jailbreakable but not available on ramdisk env.


https://man7.org/linux/man-pages/man2/pivot_root.2.html

**Reply by stephan-henningsen** [reply]
I've got the gist of it, but I'm trying to understand the nitty-gritty details.  I've briefly read the `pivot_root(2)` man page, but didn't find any obvious explanation for why `runc` doesn't work with `pivot_root` on ramdisks.

I did however find this, which gives a hint of why there might be an issue with ramdisk.  Except I don't understand what this has to do with ramdisks rather than minimalistic Linux systems that don't fully mount `procfs` and `sysfs`: https://github.com/opencontainers/runc/pull/1962#issue-398455186.

**Reply by AkihiroSuda** [reply, marked as answer]
https://man7.org/linux/man-pages/man2/pivot_root.2.html says " The rootfs (initial ramfs) cannot be `pivot_root`()ed.  "
